### PR TITLE
[INFRA] use tributors to list contributors and CITATION.cff for referencing

### LINKED ISSUE #66: [DISCUSSION] acknowledge contributors in DOIed specs on Zenodo
To facilitate citation, we should auto-generate Zenodo DOIs for each spec, with author list extracted from the contributors list in the docs.

---

edit by @sappelhoff  2020-05-06: PDF versions of the BIDS spec are now stored on zenodo: https://zenodo.org/record/3720628 ... we still have to adjust the metadata on Zenodo to properly acknowledge contributors.

### PR DESCRIPTION
**Pull request by Remi-Gau**
- closes #66
- relates to #627 

1. Add new files to becomes the single, unambiguous, authoritative representation

- the  `.tributors` file for the list of contributors, their info and their contribution type
- the  `CITATION.cff` file for the bids spec version and potentially other metada of the repo to be used when releasing on zenodo
- the `.all-contributorsrc` file for the required info to update the README contributors table

2. Add a script that uses `tools/new_contributors.tsv` as input and
 - update `.tributors`
 - uses `.tributors` to update `.all-contributorsrc` and `CITATION.CFF`

3. Add a script to update the contributors table in appendix based on `.all-contributorsrc`

Note that this table is now sorted alphabetically

4. The `.all-contributorsrc` is generated to make use all-contributors to generate a table of [contributors in the README](https://github.com/Remi-Gau/bids-specification/tree/contrib#contributors).

5. Changes little to the workflow for users to add themselves as contributors (or update their contributions): still done via the wiki, but they should now specify:

- name:
- contributions: 
- email (optional): 
- github (optional): 
- affiliation (optional): 
- orcid (optional): 
- bio (optional): 
- website (optional): 

New contributors email should in any case be provided to the maintainers privately.

6. Before a new release, maintainers

- manually update the tools/new_contributors.tsv'
- run make `update_contributors`
- a script updates:
  - `.tributors`
  - `.all-contributorsrc` (the `tributors` package handles this), 
  - `README.md` [contributor table](https://github.com/Remi-Gau/bids-specification/tree/contrib#contributors) (the `all-contributors` package handle this)
  - the authors field in `CITATION.cff` (can be scripted but may be refactored once `tributors` [can handle this](https://github.com/con/tributors/issues/58))

## TODO

- [x] identify github username of MANY contributors
- [ ] add avatar in README for contributors that have no github account
- [x] collect orcid numbers of contributors
- [x] identify eventual duplicates
- [x] ensure the final number of contributors matches the current list  
- [x] add worflow to validate CITATION.cff file
- [x] update contributing and wiki to explain how to add oneself as a contributor / update contribution lists and for maintainers how to update those files
- [ ] move institution that contributed to the acknowledgment section of the BIDS website, see also: https://github.com/bids-standard/bids-website/pull/279
- [ ] [figure out how PDFs will get released](https://github.com/bids-standard/bids-specification/pull/1115#pullrequestreview-1491846653)
- [x] [add caveat that contributors that do not add their email addresses may effectively exclude themselves from participating in elections](https://github.com/bids-standard/bids-specification/pull/1115#issuecomment-1602609124)

<hr> 

## To discuss

- [x] This will enforce an alphabetical order of contributors on zenodo. EDIT: Steering group approved of it

### CODE REVIEW

File: Makefile
Comment (yarikoptic) [inline Makefile:4]: never knew this could be done in Makefile!  I wouldn't  bother though and have a dedicated script under tools to simplify editing, avoiding double `$$` making troubleshooting more fun etc.

File: tools/tributors_to_citation.py
Comment (yarikoptic) [inline tools/tributors_to_citation.py:6]: TOO: ship along?

**Comment by yarikoptic** [comment]
> * manually update the `.tributors` file with new contributors

FWIW, here is the workflow @vsoch (welcome to provide feedback here) prepared for datalad where we get a PR if tributors finds something needing updating (e.g. a new contributor, updated metadata etc): https://github.com/datalad/datalad/blob/master/.github/workflows/update-contributors.yml  So I guess similar setup could be done here.

> identify github username of MANY contributors

FWIW, I don't think it is very reusable as-is but just an idea: here is the script I used to collate list of invitees for DataLad's JOSS paper https://github.com/datalad/datalad-git-bug-dumps/blob/master/tools/make-summaries .  The point was to invite also contributors to issues (not only code contributors), so we used  https://github.com/MichaelMure/git-bug (IIRC) to fetch all issues from datalad github repo (hence this large https://github.com/datalad/datalad-git-bug-dumps/blob/master/all-bugs.json ), and then populate list of invitees with aforementioned script.  I don't know if there are any contributors which aren't committers in BIDS though, so might not be worth the struggle.  But also it seems that git bug dump includes github logins (and names) so you could match a good portion of contributors... alternative -- just use github api to get list of all PRs and their owners, and names for those users.

File: Makefile
Comment (Remi-Gau) [reply Makefile:4]: This makefile is unlikely to survive till the end of this PR anyway.

For the ref, the idea comes from this python package cookiecutter 📦 🎂 ✂ 

https://github.com/audreyfeldroy/cookiecutter-pypackage/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/Makefile

File: tools/tributors_to_citation.py
Comment (Remi-Gau) [reply tools/tributors_to_citation.py:6]: Wanted to make sure there was some interest for the general direction of this PR before working on some final version of the script.

Don't say "I am pro open source but say I am a code exhibitionist."

**Comment by Remi-Gau** [comment]
> I don't know if there are any contributors which aren't committers in BIDS though, so might not be worth the struggle. 

In terms of issue openers that never commited I think we might have quite a few

> But also it seems that git bug dump includes github logins (and names) so you could match a good portion of contributors... alternative -- just use github api to get list of all PRs and their owners, and names for those users.

Yup that should get me quite distance already. FLW

**Review by sappelhoff — COMMENTED** [review]
Thanks @Remi-Gau -- I think this is a good initiative, also relating to some comments in #901 

Do you think we'd be able to integrate contributor email addresses here as well? AFAIK we currently only have those in a sheet shared among maintainers, because we need them for the steering group elections.

**Comment by Remi-Gau** [comment]
Thought about that. 

There are already some emails in the tributors file already that were caught  by the tributor package when it pinged the github API to get a first list of commiters.

Before adding more however, I would like to make sure that contributors are OK having their email listed on there. Would make it more convenient for us maintainers but I am not sure that all would be OK with this.

**Comment by Remi-Gau** [comment]
Question for fellow maintainers. @sappelhoff @effigies @ericearl @bendhouseart 

Currently this is shaping into having most of the info regarding contributors centralissed into a .tributor file... Except for the list of their contributions which would be in the .allcontributors

Even if the .tributor file was not made for this I am tempted to "hijack" it to put the list of contributions in there as well, so that we have only one file to update.

Any veto?

Also pinging @yarikoptic who may be able to tell if this is just a bad idea regarding the use of tributors.

**Comment by yarikoptic** [comment]
> Also pinging @yarikoptic who may be able to tell if this is just a bad idea regarding the use of tributors.

Let's ask @vsoch for the most competent advice on this. I think it should be feasible/sensible to also centralize storing types of contributions to `.tributors`.

**Comment by Remi-Gau** [comment]
> Sure what is your question? I use tributors in several of my projects if you need an example recipe.

Thanks @vsoch 

In brief: it would be easier for maintainers (and for the adventurous contributors) to only have to modify one file to update info about a contributor. We are assuming here that we are not using the contributor bot to do those updates.

At the moment, if a old contributor wants to update their list of contributions, this should be done in the `.allcontributors` file.

If a new contributor must be added, their info should go in the `.tributor` file but their list of contributions should go in the `.allcontributors` file.

Ideally I would prefer to have a single file to modify, to make easier for everyone.

So I am planning to also put the contribution list in the `.tributor` file and have a script to update everything else (`.allcontributors`, `citation.cff`, README contributors section and contributors table in the specification).

Any reason why this workflow would be a bad idea?

**Comment by vsoch** [comment]
@Remi-Gau tributors doesn't require any manual updates - the action handles that.

**Comment by Remi-Gau** [comment]
> @Remi-Gau tributors doesn't require any manual updates - the action handles that.
> 
> The point is that .tributors serves as a source of truth, and updates the other files. There isn't support for CITATION.cff (yet) but it could easily be added.

I forgot to mention that we do have contributors that are not on github and for whom we may just have a name but may have contributed to documentations, ideas in other ways (google docs...): so at some point we need to do some manual edit of some files.

The idea would be to still track their info and list of contributions in some way but to also do in one central place: hence why using `.tributors` as source of truth made sense.

But if I understand you correctly when we have a new contributor, you would suggest: add their info in their `citation.cff` (assuming tributors handles it), and also add their contributions in `allcontrib` but make sure that name of the contributor in each actually match so that tributors can properly merge that info? 

Sorry to be annoying, I may be missing something obvious about how tributor works.

**Comment by vsoch** [comment]
I think you would add to .tributors and then it would update CITATION.cff, given there is support.

**Comment by Remi-Gau** [comment]
> I think you would add to .tributors and then it would update CITATION.cff, given there is support.

okidoki

will start going down that route then, thanks. 😃

File: .tributors
Comment (Remi-Gau) [inline .tributors:60]:
```suggestion
        "blog": "http://www.chymera.eu",
        "orcid": "XXXX-XXXX-XXXX-XXXX",
```

Comment (Remi-Gau) [reply .tributors:60]: @TheChymera the orcid number you had given in the google form was missing some digits.

Do you think you could complete it here?

Thanks

Comment (ericearl) [reply .tributors:60]: @Remi-Gau I went to his paper and followed an ORCiD link: https://orcid.org/0000-0001-7037-2449

Comment (TheChymera) [reply .tributors:60]: @Remi-Gau [0000-0001-7037-2449](https://orcid.org/0000-0001-7037-2449) is indeed correct. Also my website has https — should I sumbit a PR for this?

Comment (Remi-Gau) [reply .tributors:60]: don't worry: just enough to tell me here.

Addd an "http://" to some of the "url" people put in the form because otherwise the citation.cff file is invalid according to their schema. But I forgot to check whether after doing so the URL was actually valid. 🤦🏾

**Comment by Remi-Gau** [comment]
Note to self.

- Use cff github action for validation of cff file.
- try relying on cffconvert for cff writing

**Comment by yarikoptic** [comment]
Hey @Remi-Gau are there any plans to move the functionality for the .cff generation to resolve https://github.com/con/tributors/issues/58? ;-) then it would become immediately available to other projects

**Comment by Remi-Gau** [comment]
Sort of.
I was waiting for BIDS 1.8 to be in and for the steering group elections to be over to finalize this PR.

Was planning to look into looking to see if I can use some of the code from cffconvert to avoid reinventing the wheel there.

**Comment by Remi-Gau** [comment]
@sappelhoff 
added a TODO in this PR to only keep "individuals" in the contributors and move institutions to the BIDS website.

also FYI: will work on this after the BEP for motion is in.

**Comment by Remi-Gau** [comment]
@sappelhoff 
looks like we should be able to use this even before the next release

File: CITATION.cff
Comment (Remi-Gau) [inline CITATION.cff:18]: We possibly may want to have the DOIs of the extension papers listed here

File: CITATION.cff
Comment (Remi-Gau) [inline CITATION.cff:23]: probably some better keywords can be found

**Comment by Remi-Gau** [comment]
@bids-standard/maintainers 
for anyone who may get involved in the release of the specs, I can give you a tour of what this does if you are interested (also to increase our bus factor) 🚌 🚌 🚌

File: .tributors
Comment (Remi-Gau) [inline .tributors:2937]: unsure what to do with contributors whose name we do not have

**Comment by sappelhoff** [comment]
>  for anyone who may get involved in the release of the specs, I can give you a tour of what this does if you are interested (also to increase our bus factor) 🚌 🚌 🚌

maybe a short tour during our next biweekly meeting?

**Comment by Remi-Gau** [comment]
TODO:
take into account direct changes to the contribution list: https://github.com/bids-standard/bids-specification/pull/981/files/38f38f78a1dd06f7dc5faf5c2dcc9ceeb83d2dc5..03e6badb072a005a07734440b99e7c4c1bab07b0

**Comment by Remi-Gau** [comment]
Possibly to be done after merging
- [ ] add new contributors from the new BEPS (also add them to our "internal" list of emails) to make sure they will be able to take part in coming elections

**Review by sappelhoff — APPROVED** [review]

**Review by sappelhoff — COMMENTED** [review]
discussion between @Remi-Gau and me just now:

1. the BIDS specification on zenodo is *just* the PDF, the [entry on Zenodo](https://zenodo.org/record/7263306) is created and updated manually: https://github.com/bids-standard/bids-specification/blob/8c9fa11e20d59f3ea64a3397b25cf9d45c09e3ef/Release_Protocol.md?plain=1#L217
1. **this PR** was working under the assumption that Zenodo pulls the repo contents via a webhook (on release), thereby automatically parses the newly created `CITATION.cff` file to update the zenodo metadata, and makes the PDF available with the correct metadata
1. HOWEVER, this is not what happens (see point 1)
    1. we never commit the PDF to our repo history: https://github.com/bids-standard/bids-specification/blob/8c9fa11e20d59f3ea64a3397b25cf9d45c09e3ef/Release_Protocol.md?plain=1#L142
    2. if proceeding with this PR **as is** all we would get is an archive of the bids-specification repo (**without** PDF) and the correct metadata
    3. (but we want the PDF **without** the other bids-specification repo files)

# Potential solutions

1. switch on zenodo for the bids-specification repo --> on release, an entry will be automatically created with correct metadata informed by the `CITATION.cff` from this PR
    1. in a next step, we manually upload the PDF to the other zenodo entry (as before)
    2. we copy the `CITATION.cff` metadata from the bids-specification record and add it to the manually created zenodo entry
    3. this is potentially laborious, but allows us to keep the present setup with minimal changes
    4. (we'd have to check exactly **how** laborious the copying over would be)
1. try whether dumping a `CITATION.cff` (from this repo at release time) together with a pdf into the manually curated zenodo entry can automatically updated the zenodo metadata (i.e., update the info without going via the github webhook)
    1. UPDATE: Remi tried, and this doesn't work: https://github.com/bids-standard/bids-specification/pull/1115#issuecomment-1609160740
1. create a new repository under the bids-specification org called "pdf-releases" (or similar), and hook it up to zenodo. Then start putting the BIDS spec PDFs in there piece by piece, always releasing each version. For the most recent ones, start adding the version-specific `CITATION.cff` file in addition to the version-specific PDF (for earlier versions, just upload the PDF, as we didn't have the `CITATION.cff` file back then).
    1. Potentially contact zenodo on whether they can remove / redirect the old DOIs to the new ones --> but little hope for that, see their FAQ: https://help.zenodo.org/ ... for records that have existed for a long time, modifications are unlikely to be made.
    1. if zenodo doesn't help us redirect, we will have **two** DOIs for BIDS spec PDFs from version 1.0.0 up to 1.8.0, which is unfortunate but maybe okay.

**Comment by sappelhoff** [comment]
Just a note that on the ["recent contributors" WIKI](https://github.com/bids-standard/bids-specification/wiki/Recent-Contributors), where contributors are supposed to enter their details, it currently says:

```
- name:
- contributions: 
- email (optional): 
- github (optional): 
- affiliation (optional): 
- orcid (optional): 
- bio (optional): 
- website (optional): 
```

If all details except the name is optional, we should maybe include the caveat that **contributors that we don't have contact information of are effectively excluding themselves from participating in elections** etc. -- and may not even be considered "active contributors" if we vote for a governance amendment that defines "active contributors" as those that:

1. have contributed AND
2. have not explicitly opted out AND
3. we have valid contact information of (email address)

see also:

- #1521

**Comment by Remi-Gau** [comment]
Actually I think we need to make it clear in the wiki that if people do not want to share their contact info publicly then they can do it privately but if we do not have a valid contact info then we can't count them for the votes.

**Comment by Remi-Gau** [comment]
This is in the wiki: https://github.com/bids-standard/bids-specification/wiki/Recent-Contributors#adding-yourself-as-a-contributor

> As a BIDS contributor you can get to vote during BIDS steering group elections, but the maintainers will need some contact information to reach you.
>
> If you add your email below, it will then appear on the repository (in our .tributors and CITATION.CFF file). If you do not want your email to be made public, then you should communicate your email to the maintainers privately so they can contact you.

Let me know if you think we need more.

@sappelhoff

**Comment by sappelhoff** [comment]
> Let me know if you think we need more.

I added an email they should write to (bids.maintenance+question@gmail.com, which gets redirected to anyone who has access to the account signs up for this particular "forward rule", currently only me). Other than that, it's good!

**Comment by Remi-Gau** [comment]
> 2\. ry whether dumping a CITATION CFF (from this repo at release time) together with a pdf into the manually curated zenodo entry can automatically updated the zenodo metadata (i.e., update the info without going via the github webhook)

Not seen any info in the zenodo doc that this would work.

**Comment by Remi-Gau** [comment]
> * switch on zenodo for the bids-specification repo --> on release, an entry will be automatically created with correct metadata informed by the CITATION CFF from this PR ... in a next step, when manually uploading the PDF to the other zenodo entry (as before), copy the citation cff metadata and add it to the manually created zenodo entry

Are we sure that if we do that we will keep the same generic DOI?
Should we ask zenodo in advance to see if some DOI magic needs to be done?

> * try whether dumping a CITATION CFF (from this repo at release time) together with a pdf into the manually curated zenodo entry can automatically updated the zenodo metadata (i.e., update the info without going via the github webhook)
> 
> * create a new repository under bids-specification "pdf-releases", hook it up to zenodo, and start putting the BIDS spec PDFs in there piece by piece, always releasing each version. For the most recent ones, start adding the CITATION CFFs (they were not available for the earlier versions, so think of a solution for that). Potentially contact zenodo on whether they can remove / redirect the old DOIs to the new ones.


Other possibility, github allows cutting release from any branch, so we could have a branch that we just use for realeases that has only the pdf and the citation.cff for the metadata?

**Comment by sappelhoff** [comment]
> Are we sure that if we do that we will keep the same generic DOI?
Should we ask zenodo in advance to see if some DOI magic needs to be done?

it will be a different "generic" DOI. With this solution, we'd have BIDS version PDFs <1.9 under one DOI and >= 1.9 under a different one (if one takes the "generic" DOI that zenodo creates for all versions of a particular project. each PDF will of course still have its own DOI).

But yes, perhaps we can discuss solutions with the Zenodo team, however I am afraid that most requests we could have would be rather invasive, like: "can we cancel this project and redirect all DOIs to the new one?"

> Other possibility, github allows cutting release from any branch, so we could have a branch that we just use for realeases that has only the pdf and the citation.cff for the metadata?

I would prefer having a dedicated repo for this 🤔  it seems cleaner to me and we wouldn't have to be careful if we ever wanted to archive the actual bids_specification repo content on zenodo.

---

Overall, I am in favor of my option 3 from https://github.com/bids-standard/bids-specification/pull/1115#pullrequestreview-1491846653

and in a next step asking zenodo whether they'd kindly close the currently existing zenodo project (https://zenodo.org/record/7263306) and:

1. redirect the generic DOI of that project (https://doi.org/10.5281/zenodo.3686061) to our then newly created and GitHub-linked project
2. redirect the spec version specific DOIs of that project(v1.0.0 until v1.8.0) to the then newly released spec version specific DOIs (of the newly created and GitHub-linked project)

If zenodo won't cooperate on that (which would be totally understandable, as it's a huge request) -- I'd bite the bullet of having each BIDS spec PDF from version 1.0.0 until 1.8.0 accessible under two different DOIs, which is dirty ... but life is messy. In that latter, messy, case we could upload one last release to the old zenodo project which contains just a README that says: "please look at this project instead".

**Comment by Remi-Gau** [comment]
Ok most of that makes sense to me but let's see what the other maintainers have to say about this at our next meeting

**Comment by Remi-Gau** [comment]
@sappelhoff 
most of the zenodo / pdf issues discovered via this PR are kinda independent, no?
so we could merge this PR and open an issue to start the discussion afresh (easier for other people's to follow).

**Comment by effigies** [comment]
> so we could merge this PR and open an issue to start the discussion afresh (easier for other people's to follow).

Yes. After you left the call yesterday, we agreed that it made more sense to get this in and then test any plans in the Zenodo sandbox. I don't think anybody else has the energy to review this, and it's not spec-related, it just needs to work.

**Review by effigies — APPROVED** [review]
Make buttons turn green.

**Comment by Remi-Gau** [comment]
> Make buttons turn green.

#LifeGoals

**Pull request merged**
